### PR TITLE
modules/virtualisation: add shared options, merge various diskSize options

### PR DESCRIPTION
**Pull request by phaer**
## Description of changes

There's currently a variety of `diskSize` options declared in `nixpkgs`, which all allow to specify a disk size in megabytes while some also allow a magic `auto` value.

This PR aims to start de-duplicating as many as those options as feasible. 
To do so, we introduce a new `virtualisation-options.nix` file which offers modules for shared options. While `diskSize` is the first one, it probably won't be the last. Things like memory size, desired free space in a disk image, etc could be shared in future PRs.

`virtualisation.diskSize` was already defined in `qemu-vm.nix`. Since setting this to `null` was disallowed in https://github.com/NixOS/nixpkgs/pull/338967, it's more compatible with other diskSize declaration in nixpkgs. The remaining difference is that most modules using diskSize support `auto` because they just end up passing it to `make-disk-image.nix`, but qemu and oci-image doe not. We therefore declare an internal `diskSizeAutoSupported` option which we set to false where `auto` is not supported.

I explicitly opted for an `assertion` instead of widening the enum type, as I think UX is better that way if you just set `virtualisation.diskSize` once to build several image formats and then add a module which does not support `auto`. In the latter case you'd get a type error, here you get the message of the assertion.
 
To de-duplicate those options might be a good start to standardize a few of our config options between various virtualization solutions. Doing so might hopefully lower required maintenance work and lower to barrier to build more complex functionality that works the same for different image builders & virtual machine runners.

nixfmt-rfc-style changes are all in the first commit, for easier reviews.

I built the following images succesfully but haven't been able to live test them on their respective platforms. I also haven't tested darwinBuilder yet.

* digitalOceanImage
* googleComputeImage
* linodeImage
* azureImage
* amazonImage
* openstackImage
* virtualBoxOVA
* hypervImage
* VMA # proxmoxImage
* OCIImage
* ~~brightboxImage~~ broken on master as well, see https://github.com/NixOS/nixpkgs/pull/339790

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/profiles/macos-builder.nix [hunk 1]
```diff
@@ -223,12 +248,16 @@ in
     '';
 
     virtualisation = {
-      diskSize = cfg.diskSize;
+      diskSize = lib.mkDefault 20 * 1024;
```
Comment (eclairevoyant) [inline nixos/modules/profiles/macos-builder.nix:251]:
```suggestion
      diskSize = lib.mkDefault (20 * 1024);
```

Comment (phaer) [reply nixos/modules/profiles/macos-builder.nix]: Thank you!

**Comment by phaer** [comment]
Botched up a rebase, going to fix it in a moment

**Comment by Lassulus** [comment]
Uh yeah finally someone is doing that :D It was on my TODO list since I started working on nixos-generators ;)

**Review by Lassulus — APPROVED** [review]
lgtm from a first glance

**Comment by phaer** [comment]
Rebased after brightboxImage removal got merged.

**Review by zimbatm — APPROVED** [review]

**Comment by mweinelt** [comment]
This broke all nixos tests and has as such become a channel blocker.

**Comment by mweinelt** [comment]
Reverted. Feel free to reapply after making sure the NixOS tests are in working order.

**Comment by phaer** [comment]
Thank you for the quick revert and sorry for the trouble this caused. Going to try again later and will ensure tests pass.

**Pull request merged**